File: Kamynin/1530.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
int res = 0;
int d;

vector<int> dfs(TreeNode* node){
    if (!node->left && !node->right)
        return {0};
    vector<int> left;
    vector<int> right;
    if (node->left){
        left = dfs(node->left);
        for(int i = 0; i < left.size(); ++i)
            ++left[i];
    }
    if (node->right){
        right = dfs(node->right);
        for(int i = 0; i < right.size(); ++i)
            ++right[i];
    }
    for(auto l: left)
        for(auto r: right)
            if (l + r <= d)
                ++res;
    vector<int> v(left);
    v.insert(v.end(), right.begin(), right.end());
    return v;
}

int countPairs(TreeNode* root, int distance){
    res = 0;
    d = distance;
    dfs(root);
    return res;
}
};

File: Fedorova/1530.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countPairs(self, root: TreeNode, distance: int) -> int:
        #DFS для обхода. Для потомков рекурсивно вызываем функцию в левом и правом поддеревьях
        #перебираем левый и правый списки и вычисляем количество пар,расстояния которых меньше distance.
        #возвращаем список, содержащий все расстояния от текущего узла до его потомков.
        self.ans = 0

        def dfs(root):
            if not root: return []
            if not root.left and not root.right: return [1]
            left = dfs(root.left)
            right = dfs(root.right)
            for i in left:
                for j in right:
                    if i+j <= distance:
                        self.ans += 1
            return [i+1 for i in left + right]
        dfs(root)
        return self.ans